amplipi/hw.py

```diff
@@ -290,56 +299,118 @@ def set_i2c_address(self, baud: int = 9600) -> bool:
-    """ Flash all available preamps with a given file """
+    code = subprocess.run([f'stm32flash -vb {baud} -w {filepath} {PI_SERIAL_PORT}'],
+                              shell=True, check=False).returncode
+    if code != 0:
+      # TODO: Error handling
+      print(f'Error programming {self.unit_num_to_name(unit)}, stopping programming')
+
+    # Successfully programmed unit, reset to exit bootloader
+    self.reset()
+
+    # Verify newly programmed unit communicates
+    if preamp.available():
+      print(f"{self.unit_num_to_name(unit)}'s new firmware version: {preamp.read_version()}")
+    else:
+      # Can't communicate to unit, give up
+      # TODO: retry?
+      print(f"Couldn't communicate to {self.unit_num_to_name(unit)} after programming, stopping programming")
+      return False
+
+    # Programming succeeded!
+    return True
+
+  def program_all(self, filepath: str, num_units: Union[None, int], baud: int = 115200) -> bool:
+    """ Program all available preamps with a given file
+        If num_units is not None, programming will stop after num_units
+    """
 
     if baud not in self.BAUD_RATES:
       raise ValueError(f'Baud rate must be one of {self.BAUD_RATES}')
 
-    for unit in range(num_units):
-      # If the unit was previously able to be communicated to,
-      # read the version and print it.
-      ver_str = ''
-      if unit < len(self.preamps):
-        fw_ver = self.preamps[unit].read_version()
-        ver_str = f'(version {fw_ver}) '
-      print(f"Resetting unit {unit}'s preamp {ver_str}and starting execution in bootloader ROM")
-      self.reset(unit = unit, bootloader = True)
-      for p in range(unit): # Set UART passthrough on any previous units
-        print(f'Setting unit {p} as passthrough')
-        self.preamps[p].uart_passthrough(True)
-      if unit > 0: # For now the firmware can only pass through 9600 buad to expanders
-        baud = 9600
-      flash_result = subprocess.run([f'stm32flash -vb {baud} -w {filepath} {PI_SERIAL_PORT}'], shell=True, check=False)
-      success = flash_result.returncode == 0
-      if not success:
-        # TODO: Error handling
-        print(f'Error flashing unit {unit}, stopping programming')
-      print('Resetting all preamps and starting execution in user flash')
-      self.reset()
-
-      # If the programming was successful it was just added to the list of preamps
-      if unit < len(self.preamps):
-        fw_ver = self.preamps[unit].read_version()
-        print(f"Unit {unit}'s new version: {fw_ver}")
-      elif success:
-        success = False
-        print(f"Can't communicate with unit {unit}, stopping programming")
-
-      if not success:
-        break
-    return success
+    # Force attempting to program at least 1 unit
+    if num_units is not None and num_units <= 0:
+      num_units = 1
+    previous_unit_count = len(self)
+    program_count = self.MAX_UNITS if num_units is None else num_units
```

Comment:
Just attempting to clean up and consolidate the logic here
```suggestion
    if num_units is None:
      # try to program as many units as is feasible by default
      # this is an attempt to never leave any units in a bad firmware state
      program_count = self.MAX_UNITS
    elif num_units <= 0:
      # TODO: since we were instructed to do nothing, why not just do nothing and return an error?
      return False
    else:
      program_count = num_units
    previous_unit_count = len(self)
```